proposals/module-linking/Binary.md

```diff
@@ -0,0 +1,238 @@
+# Module Linking Binary Format
+
+This document is intended to be a reference for the current state of the binary
+encoding of the module linking proposal. This is kept in sync with [the
+explainer](Explainer.md) and that is also recommended reading to understand this
+document. Currently this isn't striving to have a spec-level of detail, but
+hopefully it should be enough to get implementations off the ground and agreeing
+with each other!
+
+## Module-level updates
+
+At a high-level two new index spaces are added: modules and instances. Index
+spaces are also updated to be built incrementally as they are defined.
+
+Four new sections are added
+
+* Module section, `modulesec`, declares nested modules whose code comes
+  later. Only types are mentioned in this section.
+* Module Code section, `modulecodesec`, defines the nested modules.
+* Instance section, `instancesec`, declares and defines local instances.
+* Alias section, `aliassec`, brings in exports from nested instances or items
+  from the parent module into the local index space.
+
+The Type, Import, Module, Instance, and Alias sections can appear in any order
+at the beginning of a module. Each item defined in these sections can only refer
+to previously defined items. For example instance 4 cannot reference instance 5,
+and similarly no instance can refer to the module's own functions. Similarly
+instance 4 can only access module 3 if module 3 precedes instance 4 in binary
+format order.
+
+## Type Section updates
+
+Updates to
+[`typesec`](https://webassembly.github.io/spec/core/binary/modules.html#binary-typesec):
+
+```
+typesec ::=  t*:section_1(vec(type))
+
+type ::=
+            0x60 ft:functype                    ->        ft
+            0x61 mt:moduletype                  ->        mt
+            0x62 it:instancetype                ->        it
+
+functype ::= rt_1:resulttype rt_2:resulttype    ->        rt_1 -> rt-2
+
+moduletype ::=
+  i*:vec(import) e*:vec(exporttype)             ->        {imports i, exports e}
```

Comment:
I'm afraid this is not enough. In order to be able to type all possible modules, module types will need to be able to define types, as well as aliases. Similarly for instance types.

Reply:
Ah similar to mutually recursive types I don't think I personally know enough about type definitions/aliases in wasm to spec the binary encoding here, unfortunately :(